---
title: "Figure 2 Transcriptional changes of gene sets in the JXB (Jiaxing Black) and DLY (Duroc × Landrace × Yorkshire) pigs (N = 5 for each group). "
author: "Liang Huang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Clean workspace
rm(list=ls()) 
# Set output directory
output_dir="./"
```

#a KEGG enrichment of DEGs

(A) Bubble plot shows enriched pathways of DEGs (differentially expressed genes). 

```{r}
library(xlsx)
library(dplyr)
library(ggplot2)


file="./DEGs expression.xlsx"

down=read.xlsx(file,4,header=T)
up=read.xlsx(file,5,header=T)
all=read.xlsx(file,6,header=T)

down$Pathway <- factor(down$Pathway, levels = unique(down$Pathway))
up$Pathway <- factor(up$Pathway, levels = unique(up$Pathway))
all$Pathway <- factor(all$Pathway, levels = unique(all$Pathway))


up <- up %>%
  arrange(desc(Count))
down <- down %>%
  arrange(desc(Count))
all <- all %>%
  arrange(desc(Count))



library(ggplot2)
library(patchwork)

# 创建每个气泡图
Down <- ggplot(down, aes(x = Gene_ratio, y = reorder(Pathway, Count))) +
  geom_point(aes(size = Count, color = PValue)) +
  scale_color_gradient(low = "#FF0000", high = "#0033FF") +
  labs(y="Down",x="Gene ratio" #title = "KEGG pathway", subtitle = "(Up-regulated genes in W)"
    ) +
  theme_test() +
  theme(
    plot.title = element_text(hjust = 0.5, vjust = 0, size = 20),
    plot.subtitle = element_text(hjust = 0.5, vjust = 0, size = 15),
    axis.title = element_text(size = 20),
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    #axis.text.x = element_text(angle = -45, size = 14, vjust = 1, hjust = 0),
    text = element_text(size = 16)
  )
Down


Up <- ggplot(up, aes(x = Gene_ratio, y = reorder(Pathway, Count))) +
  geom_point(aes(size = Count, color = PValue)) +
  scale_color_gradient(low = "#FF0000", high = "#0033FF") +
  labs(y="Up" #title = "KEGG pathway", subtitle = "(Up-regulated genes in B)"
       ) +
  theme_test() +
  theme(
    plot.title = element_text(hjust = 0.5, vjust = 0, size = 20),
    plot.subtitle = element_text(hjust = 0.5, vjust = 0, size = 15),
    axis.title = element_text(size = 20),
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    #axis.text.x = element_text(angle = -45, size = 14, vjust = 1, hjust = 0),
    text = element_text(size = 16)
  )
Up

library(ggplot2)
All <- ggplot(all, aes(x = Gene_ratio, y = reorder(Pathway, Count))) +
  geom_point(aes(size = Count , color = PValue)) +
  scale_color_gradient(low = "#FF0000", high = "#0033FF") +
  #labs(y="All" #title = "KEGG pathway", subtitle = "(All differently expressed genes)"
  #     ) +
  theme_test() +
  theme(
    plot.title = element_text(hjust = 0.5, vjust = 0, size = 20),
    plot.subtitle = element_text(hjust = 0.5, vjust = 0, size = 15),
    axis.title = element_text(size = 12),
    axis.title.y = element_blank(),
    #axis.title.x = element_blank(),
    #axis.text.x = element_text(angle = -45, size = 14, vjust = 1, hjust = 0),
    text = element_text(size = 12)
  )+labs(x="RichFactor",size="Count")

All

combined_plot <- Up +
  plot_layout(ncol = 1, heights = c(0.4, 0.4)) +
  Down +
  plot_layout(ncol = 1, heights = 0.4) +
  All +
  plot_layout(ncol = 1)

combined_plot

#plot were adjusted using Adobe Illustrator 
```



#b heatmap

(B) Heatmap displayed expression of DEGs, pathways they belong to is also marked.

```{r}
data=read.xlsx(file,header=T, 1)
group=read.xlsx(file,header=T, 2,row.names=1)

library(ComplexHeatmap)

annotation_row=data[,22,drop=F]
rownames(annotation_row)=data[,2]

df=data[,c(10:19)]
rownames(df)=data[,2]


ann_colors = list(
  group = c(DLY = "#1B9E77", JXB = "#D95F02"))

pheatmap(df, annotation_col = group, annotation_row = annotation_row, 
         annotation_colors = ann_colors,
         show_rownames = F,
         show_colnames = TRUE,
         scale="row")

#Heatmap were adjusted using Adobe Illustrator 

```



#c-f WGCNA

(C) The cluster dendrogram. (D) Gene count of each module. (E) Cladogram of modules. (F) Correlations between species and modules by WGCNA (Weighted Gene Co-expression Network Analysis).Different colors represent different modules that consist of genes with similar expression mode, p value were displayed below correlation coefficient.

```{r}
exprmat.quantiles=read.xlsx(file,header = T,1,row.names=2)
exprmat.quantiles=exprmat.quantiles[,c(9:18)]

group=read.xlsx(file,header = T,2,row.names=1)
trait=read.xlsx(file,header = T,7,row.names=1)


library(WGCNA)
enableWGCNAThreads(6)
options(stringsAsFactors = FALSE);

dat0=exprmat.quantiles


Exp <- factor(group$group)
design <- model.matrix(~ 0+Exp)

datSummary=dat0[,1];

dim(dat0) 


datExpr = t(dat0[,1: ncol(dat0)]);
dim(datExpr)

ArrayName= names(data.frame(dat0)) 

GeneName= dat0[,1]

names(datExpr)=dat0[,1]

y=design[,2]


setwd("./WGCNA") 
#####WGCNA#####

powers=c(seq(1,10,by=1),seq(12,16,by=2));

sft=pickSoftThreshold(datExpr, powerVector=powers,networkType = "signed")


sizeGrWindow(9, 5); 

pdf('choosing power.pdf');

par(mfrow = c(1,1)); 

cex1 = 0.9;

plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",main = paste("Scale independence")); 

text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],labels=powers,cex=cex1,col="red");

abline(h=0.90,col="red"); 

dev.off() 


sizeGrWindow(9, 5);

pdf('mean connectivity.pdf');

plot(sft$fitIndices[,1], sft$fitIndices[,5],xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",main = paste("Mean connectivity"))

text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red");

dev.off()



#####WGCNA#####

softPower =16

Connectivity=softConnectivity(datExpr,corFnc = "cor", corOptions = "use ='p'",power=softPower,type = "signed")

pdf("scale-free.pdf");

scaleFreePlot(Connectivity,nBreaks = 10,truncated = FALSE,removeFirst = FALSE, main = "");

dev.off()

adjacency = adjacency(datExpr,corFnc = "cor", corOptions = "use ='p'",type = "signed", power = softPower)

TOM = TOMsimilarity(adjacency,TOMType="signed");

dissTOM = 1-TOM

geneTree = hclust(as.dist(dissTOM), method = "average")

minModuleSize =10; 

dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM,deepSplit = 4, pamRespectsDendro = FALSE,minClusterSize = minModuleSize,cutHeight=0.99); #deepSplit = 0-4

table(dynamicMods)

dynamicColors = labels2colors(dynamicMods)

table(dynamicColors)

MEList = moduleEigengenes(datExpr, colors = dynamicMods) #colors = dynamicColors

MEs = MEList$eigengenes

MEDiss = 1-cor(MEs);

METree = hclust(as.dist(MEDiss), method = "average");

sizeGrWindow(7, 6)

plot(METree, main = "Clustering of module eigengenes",xlab = "", sub = "")

MEDissThres = 0.2

abline(h=MEDissThres, col = "red")

merge = mergeCloseModules(datExpr, dynamicMods, cutHeight = MEDissThres, verbose = 3); 

mergedColors = merge$colors;

table(mergedColors)

mergedMEs = merge$newMEs;

sizeGrWindow(12, 9)

pdf("DendroAndColors.pdf")

plotDendroAndColors(geneTree, cbind(dynamicMods, mergedColors),c("Dynamic Tree Cut", "Merged dynamic"),dendroLabels = FALSE, hang = 0.03,addGuide = TRUE, guideHang = 0.05)

dev.off()



moduleColors = mergedColors

MEs = mergedMEs;

MEDiss = 1-cor(MEs);

METree = hclust(as.dist(MEDiss), method = "average");

pdf("METree.pdf")

plot(METree, main = "Clustering of module eigengenes",xlab = "", sub = "")

dev.off()

nSamples=nrow(datExpr)

geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p")); 

MMPvalue = cbind.data.frame(datSummary,corPvalueStudent(as.matrix(geneModuleMembership), nSamples)); 

write.table(data.frame(ArrayName,MEs),"MEs.csv",row.name=F) 

kMEdat=data.frame(geneModuleMembership,MMPvalue)

write.table(data.frame(datSummary,kMEdat),"kME-MMPvalue.csv",row.names=FALSE)

k.in=intramodularConnectivity(adjacency(datExpr,corFnc = "cor", corOptions = "use ='p'", type = "signed", power = softPower), moduleColors,scaleByMax = FALSE)

datout=data.frame(datSummary, colorNEW=moduleColors, k.in)

write.table(datout, file="OutputCancerNetwork.csv", sep=",", row.names=F) 

hubs= chooseTopHubInEachModule(datExpr, moduleColors)

write.csv(data.frame(module=names(hubs),moduleColor=labels2colors(names(hubs)),hub=hubs),"num2color.csv",row.names=F)

#####correlation between gene expression and trait#####


signif(cor(y,MEs, use="p"),2)

p.values = corPvalueStudent(cor(y,MEs, use="p"), nSamples = nSamples) #trait和模块相关分析的P值

#Measure of module significance as average gene significance

GS1=as.numeric(cor(y,datExpr, use="p")) #trait和基因表达的相关系数

GeneSignificance=abs(GS1)

# Next, module significance is defined as average gene significance.

ModuleSignificance=tapply(GeneSignificance, moduleColors, mean, na.rm=T) #求模块内GS1的均值

sizeGrWindow(8,7)

par(mfrow = c(1,1))

plotModuleSignificance(GeneSignificance[moduleColors!=0],moduleColors[moduleColors!=0])

#plot Gene significance (y-axis) vs. intramodular connectivity (x-axis) 画图gene sig和连接度散点

colorlevels=unique(moduleColors)

colorlevels=colorlevels[colorlevels!=0]

sizeGrWindow(9,6)

par(mfrow=c(2,as.integer(0.5+length(colorlevels)/2)))

par(mar = c(4,5,3,1))

for (i in c(1:length(colorlevels)))

{

  whichmodule=colorlevels[[i]];

  restrict1 = (moduleColors==whichmodule);

  verboseScatterplot(k.in$kWithin[restrict1],

                     GeneSignificance[restrict1], col=moduleColors[restrict1],

                     main=whichmodule,

                     xlab = "Connectivity", ylab = "Gene Significance", abline = TRUE)

}
# 
#Generalizing intramodular connectivity for all genes on the array

datKME=signedKME(datExpr, MEs, outputColumnName="MM.")

# Display the first few rows of the data frame

head(datKME)



#####Trait and gene cor#####

NS1=networkScreening(y=y, datME=MEs, datExpr=datExpr,oddPower=softPower, blockSize=10000, minimumSampleSize=4,addMEy=TRUE, removeDiag=FALSE, weightESy=0.5)

GeneResultsNetworkScreening=data.frame(GeneName=row.names(NS1), NS1)

#write.table(GeneResultsNetworkScreening, file="GeneResultsNetworkScreening.csv",row.names=F,sep=",")

MEsy = data.frame(y, MEs)

eigengeneSignificance = cor(MEsy, y); 

eigengeneSignificance[1,1] = (1+max(eigengeneSignificance[-1, 1]))/2

eigengeneSignificance.pvalue = corPvalueStudent(eigengeneSignificance, nSamples = length(y)) 

namesME=names(MEsy)

# Form a summary data frame

out1=data.frame(t(data.frame(eigengeneSignificance,eigengeneSignificance.pvalue, namesME, t(MEsy))))

# Set appropriate row names

dimnames(out1)[[1]][1]="EigengeneSignificance"

dimnames(out1)[[1]][2]="EigengeneSignificancePvalue"

dimnames(out1)[[1]][3]="ModuleEigengeneName"

dimnames(out1)[[1]][-c(1:3)]=dimnames(datExpr)[[1]]

# Write the data frame into a file

write.table(out1, file="MEResultsNetworkScreening.csv", row.names=TRUE, col.names = TRUE, sep=",")

GeneName=dimnames(datExpr)[[2]]

GeneSummary=data.frame(GeneName, moduleColors, NS1)

write.table(GeneSummary, file="GeneSummaryTutorial.csv", row.names=F,sep=",")

datTraits=data.frame(ArrayName, MEsy)

dimnames(datTraits)[[2]][2:length(namesME)]=paste("Trait",dimnames(datTraits)[[2]][2:length(namesME)],sep=".")

write.table(datTraits, file="TraitsTutorial.csv", row.names=F,sep=",")

#Relationships among the top 30 most significant genes,correlation heatmaps for signed network:

sizeGrWindow(7,7)

NS1=networkScreening(y=y, datME=MEs, datExpr=datExpr,oddPower=softPower, blockSize=10000, minimumSampleSize=4,addMEy=TRUE, removeDiag=FALSE, weightESy=0.5)

topList=rank(NS1$p.Weighted,ties.method="first")<=30

#gene.names= names(datExpr)[topList]

gene.names=GeneName[topList]

colnames(datExpr)=GeneName

# The following shows the correlations between the top genes

plotNetworkHeatmap(datExpr, plotGenes = gene.names,networkType="signed", useTOM=TRUE,power=softPower, main="signed correlations")





#####automatic finish the Cytoscape mods#####

probes = dat0[,1]

n=length(unique(moduleColors)[unique(moduleColors)!=0])

pb <- txtProgressBar(min = 0, max = n, style = 3)#添加进度条

for (p in 1:n) { modules=unique(moduleColors)[unique(moduleColors)!=0][p]

inModule = is.finite(match(moduleColors,modules));

modProbes = probes[inModule];

modTOM = TOM[inModule, inModule];

dimnames(modTOM) = list(modProbes, modProbes)

cyt = exportNetworkToCytoscape(modTOM,

                               edgeFile = paste("CytoscapeInput-edges-", paste(modules, collapse="-"), ".txt", sep=""),

                               nodeFile = paste("CytoscapeInput-nodes-", paste(modules, collapse="-"), ".txt", sep=""),

                               weighted = TRUE,threshold = quantile(abs(modTOM),probs=0.8,nodeNames = modProbes ,nodeAttr = moduleColors[inModule]))

#threshold can be replaced by quantile(abs(modTOM),probs=0.8)

setTxtProgressBar(pb, p)}

close(pb)

```



#g network of genes in module

(G) Network showed key driver of modules, gradient color and size of nodes represent degree (edge count of genes), red indicates high degree and blue indicates low degree.

#These figure were drawn by using Cytoscape.

